sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeRowsExec.scala

```diff
@@ -0,0 +1,216 @@
+    def projectExtraRow(row: InternalRow): InternalRow = otherProjection.apply(row)
+  }
+
+  sealed trait CardinalityValidator {
+    def validate(row: InternalRow): Unit
+  }
+
+  object NoopCardinalityValidator extends CardinalityValidator {
+    def validate(row: InternalRow): Unit = {}
+  }
+
+  /**
+   * A simple cardinality validator that keeps track of seen row IDs in a roaring bitmap.
+   * This validator assumes the target table is never broadcasted or replicated, which guarantees
+   * matches for one target row are always co-located in the same partition.
+   *
+   * IDs are generated by [[org.apache.spark.sql.catalyst.expressions.MonotonicallyIncreasingID]].
+   */
+  case class BitmapCardinalityValidator(rowIdOrdinal: Int) extends CardinalityValidator {
+    private val matchedRowIds = new Roaring64Bitmap()
+
+    override def validate(row: InternalRow): Unit = {
+      val currentRowId = row.getLong(rowIdOrdinal)
+      if (matchedRowIds.contains(currentRowId)) {
+        throw QueryExecutionErrors.mergeCardinalityViolationError()
+      }
+      matchedRowIds.add(currentRowId)
+    }
+  }
+
+  /**
+   * An iterator that acts on joined target and source rows and computes deletes, updates and
+   * inserts according to provided MERGE instructions.
+   *
+   * If a particular joined row should be discarded, this iterator returns null.
+   */
+  class MergeRowIterator(
+      private val rowIterator: Iterator[InternalRow],
+      private val cardinalityValidator: CardinalityValidator,
+      private val isTargetRowPresentPred: BasePredicate,
+      private val isSourceRowPresentPred: BasePredicate,
+      private val matchedInstructions: Seq[InstructionExec],
+      private val notMatchedInstructions: Seq[InstructionExec],
+      private val notMatchedBySourceInstructions: Seq[InstructionExec])
+    extends Iterator[InternalRow] {
+
+    var cachedExtraRow: InternalRow = _
+
+    override def hasNext: Boolean = cachedExtraRow != null || rowIterator.hasNext
+
+    override def next(): InternalRow = {
+      if (cachedExtraRow != null) {
+        val extraRow = cachedExtraRow
+        cachedExtraRow = null
+        return extraRow
+      }
+
+      val row = rowIterator.next()
+
+      val isSourceRowPresent = isSourceRowPresentPred.eval(row)
+      val isTargetRowPresent = isTargetRowPresentPred.eval(row)
+
+      if (isTargetRowPresent && isSourceRowPresent) {
+        cardinalityValidator.validate(row)
+        applyInstructions(row, matchedInstructions)
+      } else if (isSourceRowPresent) {
+        applyInstructions(row, notMatchedInstructions)
+      } else if (isTargetRowPresent) {
+        applyInstructions(row, notMatchedBySourceInstructions)
+      } else {
+        null
+      }
+    }
+
+    private def applyInstructions(
+        row: InternalRow,
+        instructions: Seq[InstructionExec]): InternalRow = {
+
+      for (instruction <- instructions) {
+        if (instruction.condition.eval(row)) {
```

Comment:
Not using `find` to avoid extra calls as it is called for every row.

Reply:
Regarding https://github.com/databricks/scala-style-guide#traversal-and-zipwithindex

`while` is preferred, I'm not sure if it still suitable for the current Scala compiler

Reply:
Let me dig into the generated bytecode while adding a benchmark as part of SPARK-44013.

Reply:
OK, thanks, @aokolnychyi